src/lib.rs

```diff
@@ -251,19 +266,135 @@ pub mod test {
     }
 
     #[test]
-    fn forth() {
+    fn sizes() {
         use core::mem::{align_of, size_of};
         assert_eq!(5 * size_of::<usize>(), size_of::<DictionaryEntry<()>>());
         assert_eq!(5 * size_of::<usize>(), size_of::<DictionaryEntry<()>>());
         assert_eq!(1 * size_of::<usize>(), align_of::<Word>());
+    }
 
+    #[test]
+    fn forth() {
         let mut lbforth = LBForth::from_params(
             LBForthParams::default(),
             TestContext::default(),
             Forth::<TestContext>::FULL_BUILTINS,
         );
+
+        test_forth(&mut lbforth.forth,|forth| forth.process_line(), |forth| forth);
+
+        let context = lbforth.forth.release();
+        assert_eq!(&context.contents, &[6, 5, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9]);
+    }
+
+    struct CountingFut<'forth> {
+        target: usize,
+        ctr: usize,
+        forth: &'forth mut Forth<TestContext>,
+    }
+
+    impl<'forth> Future for CountingFut<'forth> {
+        type Output = Result<(), Error>;
+
+        fn poll(mut self: core::pin::Pin<&mut Self>, cx: &mut core::task::Context<'_>) -> core::task::Poll<Self::Output> {
+            match self.ctr.cmp(&self.target) {
+                Ordering::Less => {
+                    self.ctr += 1;
+                    cx.waker().wake_by_ref();
+                    Poll::Pending
+                },
+                Ordering::Equal => {
+                    self.ctr += 1;
+                    let word = Word::data(self.ctr as i32);
+                    self.forth.data_stack.push(word)?;
+                    Poll::Ready(Ok(()))
+                },
+                Ordering::Greater => {
+                    Poll::Ready(Err(Error::InternalError))
+                },
+            }
+        }
+    }
+
+    #[cfg(feature = "async")]
+    #[test]
+    fn async_forth() {
+        use crate::{dictionary::{AsyncBuiltins, AsyncBuiltinEntry}, fastr::FaStr, async_builtin, leakbox::AsyncLBForth};
+
+        struct TestAsyncDispatcher;
+        impl<'forth> AsyncBuiltins<'forth, TestContext> for TestAsyncDispatcher {
+            type Future = CountingFut<'forth>;
```

Comment:
@hawkw how does this work when you have more than one word you support? Do you end up needing to manually make an enum of all the different potential futures that could go here?

If so, is there any way to *not* have that, other than "async method in trait"?

Reply:
@jamesmunns I actually discussed this a bit in the documentation here: https://github.com/jamesmunns/forth3/blob/45228ae57d206542b4df932be01bc7c602c776cd/src/dictionary.rs#L102-L120

The future returned by `dispatch_async` can either be:
-  an `enum` of all the potential builtin word futures
- a `Pin<Box<dyn Future<Output = Result<(), Error>>>` (when `alloc` is available)
-  an `impl Future<Output = Result<(), Error>>` (when nightly features are available)
 Each of these approaches has different tradeoffs:
- the `enum` requires all futures to be named (they can't be `async fn`s) and must be implemented manually, which is annoying
- the boxed `dyn Future` requires an allocation to erase the future type,
- `impl Future` requires the `#![feature(impl_trait_in_assoc_type)]` and thus nightly Rust.

There's no way to erase the future types without requiring one of those approaches, and _not_ having some way to erase the future types is impractical, as it would force the entire `AsyncForth` type to be generic over each possible future type. But, the approach of using a trait with an associated type for `AsyncBuiltins::Future` at least allows the user code to decide which of those tradeoffs is best suited for the application's use case.

IMO, in our OS, we should use the `impl Future` approach, as we won't be using liballoc and can't use `Box`, and I think it's acceptable to build the top-level OS binary on nightly to get access to `impl Trait` in associated types.

Reply:
If you like, I can try to add a test or example that shows the various approaches for erasing multiple future types.